app/models/user.rb

```diff
@@ -0,0 +1,56 @@
+# == Schema Information
+#
+# Table name: users
+#
+#  id              :integer          not null, primary key
+#  username        :string           not null
+#  password_digest :string           not null
+#  session_token   :string           not null
+#  created_at      :datetime         not null
+#  updated_at      :datetime         not null
+#
+
+class User < ApplicationRecord
+  validates :username, :password_digest, :session_token, presence: true
+  validates :username, uniqueness: true
+  validates :password, length: { minimum: 6, allow_nil: true }
```

Comment:
I'm surprised by the `allow_nil: true` here. Why was that added?

Reply:
In my auth pattern, I am not storing the password in my db, but rather the hashed password leveraging BCrypt.  The @password attribute is only set if a user creates a new password with the #password= method.  That method, hashes the password and saves the hashed password in the db.  However, when we fetch this user at another time, since the password was only an instance variable and didn't get saved in the db, the password will be nil.  If I didn't have the allow_nil: true validation, then every user would be marked as invalid since the password (nil) would not meet the length requirement.  In reality, we only need to validate the password when it has been changed, which is exactly when @password is not nil, however the allow_nil true removes the chance for the bug I mentioned above.